File: petsc/pattern.py
class Pattern:
    def __init__(self, pattern, projection, candidates):
        self.pattern = pattern
        self.projection = projection
        self.candidates = candidates

        # When instantiated, Pattern is added to queue, which is a max-heap
        self.support = -len(projection)

        # Ability to reproduce this pattern from original data
        self.signal = None
        self.window = None

    def __repr__(self):
        return f"Pattern({self.pattern}, {self.support})"

    def __len__(self):
        return len(self.pattern)

    def __eq__(self, other):
        if not isinstance(other, Pattern):
            return NotImplemented
        return self.support == other.support

    def __lt__(self, other):
        if not isinstance(other, Pattern):
            return NotImplemented
        return self.support < other.support

File: src/pattern.cpp
#include "pattern.h"

Pattern::Pattern(Word pattern, Projection projection,
                 Candidates candidates, int support)
    : pattern(std::move(pattern)), projection(std::move(projection)),
      candidates(std::move(candidates)), support(support) {
  coef = 0.0;
}

bool Pattern::operator<(const Pattern &other) const {
  return support < other.support;
}

bool Pattern::operator>(const Pattern &other) const {
  return support > other.support;
}
